SDWebImage/SDWebImageDownloaderOperation.m

```diff
@@ -262,11 +250,26 @@ - (void)reset {
     SD_LOCK(self.callbacksLock);
     [self.callbackBlocks removeAllObjects];
     SD_UNLOCK(self.callbacksLock);
-    self.dataTask = nil;
     
-    if (self.ownedSession) {
-        [self.ownedSession invalidateAndCancel];
-        self.ownedSession = nil;
+    @synchronized (self) {
```

Comment:
？What the reason to use `@ synchronized`. The `self.dataTask` and `self.ownedSession` is immutable after we initilize in in the start method. No any write operation to that operation.

Reply:
@dreampiggy `reset` method can be called concurrently, for `dataTask`, read operation like https://github.com/SDWebImage/SDWebImage/blob/e79f35f0849755d3d2db8e4429f955f027ed2c69/SDWebImage/SDWebImageDownloaderOperation.m#L227-L228, set operation like https://github.com/SDWebImage/SDWebImage/blob/e79f35f0849755d3d2db8e4429f955f027ed2c69/SDWebImage/SDWebImageDownloaderOperation.m#L255
https://github.com/SDWebImage/SDWebImage/blob/e79f35f0849755d3d2db8e4429f955f027ed2c69/SDWebImage/SDWebImageDownloaderOperation.m#L403
So we can't guarantee dataTask would not be nil when executing `[self.dataTask cancel]`.

Reply:
Check you logic again. I remember there are previous a crash because of `self.imageData = nil` (trigger from user's `[operation cancel]` queue) can happend when `URLSessionDelegate` call `didReceiveData:` (triggered in URLSession delegate queue) at the same time.

However, the`dataTask.cancel` is different. If you walkthrough this logic, there should not be chance to trigger `self.dataTask = nil` && `self.dataTask` at the same time. Because this is a `NSOperation`, the cancel state will cause the operation queue remove it. While the URLSession will also stop further delegate callback.

Reply:
No, operation queue remove operation only if it's not started, that is `executing` is `NO`. session delegate queue, coder queue, both can call `done` method, and any other thread can be called when user call `cancel`. Think about thread context switch and parallel, concurrency, I don't know what you said no chance.

Reply:
Give your code path. I rember at that time (Half year ago) I check the total path for that any write operation insdie `reset` method. Other property does not cause thread-safe issue.

Reply:
User call `cancel` and url session queue call `done` can enter `reset` method concurrently, right? If it is, can concurrently execute code below, right ? 
```
    self.dataTask = nil;
    
    if (self.ownedSession) {
        [self.ownedSession invalidateAndCancel];
        self.ownedSession = nil;
    }
```